components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java

```diff
@@ -4274,19 +4322,31 @@ private void deleteAPI(API api) throws APIManagementException {
         // Get the URI Templates for the given API to detach the resources scopes from
         Set<URITemplate> uriTemplates = apiMgtDAO.getURITemplatesOfAPI(apiIdentifier);
         // Detach all the resource scopes from the API resources in KM
-        KeyManagerHolder.getKeyManagerInstance(tenantDomain).detachResourceScopes(api, uriTemplates);
-        if (log.isDebugEnabled()) {
-            log.debug("Resource scopes are successfully detached for the API : " + apiIdentifier
-                    + " from KeyManager.");
-        }
-        // remove the local scopes from the KM
-        for (String localScope : localScopeKeysToDelete) {
-            KeyManagerHolder.getKeyManagerInstance(tenantDomain).deleteScope(localScope);
-        }
-        if (log.isDebugEnabled()) {
-            log.debug("Local scopes are successfully deleted for the API : " + apiIdentifier
-                    + " from KeyManager.");
+        Map<String, KeyManagerDto> tenantKeyManagers = KeyManagerHolder.getTenantKeyManagers(tenantDomain);
+        for (Map.Entry<String, KeyManagerDto> keyManagerDtoEntry : tenantKeyManagers.entrySet()) {
+            KeyManager keyManager = keyManagerDtoEntry.getValue().getKeyManager();
+            if (keyManager != null) {
+                try {
+                    keyManager.detachResourceScopes(api, uriTemplates);
+                    if (log.isDebugEnabled()) {
+                        log.debug("Resource scopes are successfully detached for the API : " + apiIdentifier
+                                + " from KeyManager :" + keyManagerDtoEntry.getKey() + ".");
+                    }
+                    // remove the local scopes from the KM
+                    for (String localScope : localScopeKeysToDelete) {
+                        keyManager.deleteScope(localScope);
+                    }
+                    if (log.isDebugEnabled()) {
+                        log.debug("Local scopes are successfully deleted for the API : " + apiIdentifier
+                                + " from KeyManager : " + keyManagerDtoEntry.getKey() + ".");
+                    }
+                } catch (APIManagementException e) {
+                    log.error("Error while Detach and Delete Scope from KeyManager " + keyManagerDtoEntry.getKey());
+                }
+            }
         }
+
```

Comment:
shouldn't we also delete the local scopes from AM_SCOPES ?

```suggestion
deleteScopes(localScopeKeysToDelete, tenantId);
```